File: Day-11: Greedy Algorithm/5) Coin Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int n = coins.length;
        int[][] dp = new int[n+1][amount+1];
        for(int i = 0; i < n+1; i++)
        {
            for(int j = 0; j < amount+1; j++)
            {
                dp[i][j] = -1;
            }
        }
        int ans = helper(coins, amount, 0, dp);
        if(ans == Integer.MAX_VALUE - 1)
        {
            return -1;
        }
        return ans;
    }
    public int helper(int[] coins, int amount, int id, int[][] dp)
    {
        int n = coins.length;
        if(id >= n || amount <= 0)
        {
            if(amount == 0)
            {
                return 0;
            }
            return Integer.MAX_VALUE-1;
        }
        if(dp[id][amount] != -1)
        {
            return dp[id][amount];
        }
        if(coins[id] > amount)
        {
            dp[id][amount] = helper(coins, amount, id+1, dp);
        }
        else
        {
            int take = 1 + helper(coins, amount - coins[id], id, dp);
            int notake = helper(coins, amount, id+1, dp);
            dp[id][amount] = Math.min(take, notake);
        }
        return dp[id][amount];
    }
}

File: Day-11: Greedy Algorithm/5) Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n+1, vector<int>(amount+1, -1));
        int ans =  helper(0, coins, amount, dp);
        if(ans == INT_MAX - 1)
        {
            return -1;
        }
        return ans;
    }
    int helper(int id, vector<int> &coins, int amount, vector<vector<int>> &dp)
    {
        int n = coins.size();
        if(id >= n || amount <= 0)
        {
            if(amount == 0)
            {
                return 0;
            }
            return INT_MAX-1;
        }
        if(dp[id][amount] != -1)
        {
            return dp[id][amount];
        }
        if(coins[id] > amount)
        {
            int notake = helper(id+1, coins, amount, dp);
            dp[id][amount] = notake;
        }
        else
        {
            int take = 1 + helper(id, coins, amount - coins[id], dp);
            int notake = helper(id+1, coins, amount, dp);
            dp[id][amount] = min(take, notake);
        }
        return dp[id][amount];
    }
};
